class6.1/main.cpp

```diff
@@ -0,0 +1,187 @@
+#include <iostream>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <functional>
+#include <vector>
+#include <random>
+#include <chrono>
+
+
+struct GreenMessage {
+    std::string message;
+    int counter;
+};
+
+struct BlueMessage {
+    double value1;
+    double value2;
+};
+
+struct OrangeMessage {
+    std::string str1;
+    std::string str2;
+    int number;
+    double value;
+};
+
+
+class MessageSubscriber {
+public:
+    virtual ~MessageSubscriber() {}
+    virtual void handleMessage(const GreenMessage& msg) = 0;
+    virtual void handleMessage(const BlueMessage& msg) = 0;
+    virtual void handleMessage(const OrangeMessage& msg) = 0;
+};
+
+
+class MessageDispatcher {
+public:
+
+    void subscribe(MessageSubscriber* subscriber) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        subscribers_.push_back(subscriber);
+    }
+
+
+    void publish(const GreenMessage& message) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        for (auto subscriber : subscribers_) {
+            subscriber->handleMessage(message);
+        }
+    }
+
+
+    void publish(const BlueMessage& message) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        for (auto subscriber : subscribers_) {
+            subscriber->handleMessage(message);
```

Comment:
In this case it is not a good idea to call callbacks from subscribers as it may hang the thread of the caller for some time(if one of the callbacks is big and complex enough) who just try to publish some message instantly without waiting of everyone to process the message.
There is another indicator that something is wrong with current implementation - the function name doesn't match what the function really does - now it is more publishMessageAndHandleIt not publish.
Therefore it is better to have a separate thread that will perform dispatching of the messages. As a result publisher will be able to publish message almost instantly by just pushing a message to the queue and after that he will be able to continue his regular work. While callbacks will be sent to subscribers async in other thread.